ldap/servers/plugins/replication/cl5_api.c

```diff
@@ -1705,6 +1739,227 @@ cl5DBData2Entry(const char *data, PRUint32 len __attribute__((unused)), CL5Entry
+             * of the transaction.
+             */
+            rc = TXN_BEGIN(cldb, NULL, &txnid, readonly);
+            if (rc != 0) {
+                rc = _cl5Dberror(cldb, rc, "_cl5Iterate - Failed to begin transaction");
+                continue;
+            }
+        } else {
+            /* read-only opertion on bdb are transactionless, so no reason to abort txn 
+             * after having seen some number of records 
+             */
+            dblcictx->seen.nbmax = 0;
+        }
+        /* create cursor */
+        rc = dblayer_new_cursor(cldb->be, cldb->db, txnid, &dblcictx->cursor);
+        if (rc != 0) {
+            rc = _cl5Dberror(cldb, rc, "_cl5Iterate - Failed to create cursor");
+            TXN_ABORT(cldb, txnid);
+            continue;
+        }
+        if (startcsn) {
+            char csnstr[CSN_STRSIZE] = "";
+            dbi_val_t startingkey = {0};
+            csn_as_string(startcsn, 0, csnstr);
+            dblayer_value_set(cldb->be, &startingkey, csnstr, CSN_STRSIZE);
+            rc = dblayer_cursor_iterate(&dblcictx->cursor, action_cb, &startingkey, dblcictx);
+        } else {
+            rc = dblayer_cursor_iterate(&dblcictx->cursor, action_cb, NULL, dblcictx);
+        }
+        if (dblcictx->seen.nbmax > 10 && rc == CL5_DB_LOCK_ERROR) {
+            /*
+             * Ran out of locks, need to restart the transaction.
+             * Reduce the the batch count and reset the key to
+             * the starting point.
+             * PR: Not sure weither this code is still working because dbimpl API layer
+             *  may remap bdb error 12 to something else.
+             */
+            dblcictx->seen.nbmax -= 10;
+            slapi_log_err(SLAPI_LOG_REPL, repl_plugin_name_cl,
+                          "_cl5Iterate - Ran out of db locks while iterating on entries; "
+                          "Reducing the batch value to %d and retry.\n", dblcictx->seen.nbmax);
+            nbtries = 0;
+            continue;
+        }
+        rc = _cl5Dberror(cldb, rc, NULL);
+        if (rc != CL5_NOTFOUND && rc != CL5_SUCCESS) {
+            dblcictx->changed.nb = 0; /* Tells to abort the txn */
+            dblcictx->seen.nb = 0;
+        }
+        dblayer_cursor_op(&dblcictx->cursor, DBI_OP_CLOSE, NULL, NULL);
+        dblcictx->changed.tot += dblcictx->changed.nb;
+        dblcictx->seen.tot += dblcictx->seen.nb;
+
+        if (txnid) {
+            if (dblcictx->changed.nb) {
+                int rc2 = TXN_COMMIT(cldb, txnid);
+                if (rc2 != DBI_RC_SUCCESS) {
+                    rc = _cl5Dberror(cldb, rc2, "_cl5Iterate - Failed to commit transaction");
+                }
+            } else {
+                int rc2 = TXN_ABORT(cldb, txnid);
+                if (rc2 != DBI_RC_SUCCESS) {
+                    rc = _cl5Dberror(cldb, rc2, "_cl5Iterate - Failed to abort transaction");
+                    if (rc == CL5_SUCCESS || rc == CL5_NOTFOUND) {
+                        rc = rc2;
```

Comment:
I wonder if this line is correct (I know it is cut/paste old code). Here _cl5Dberror translate a DBI error (rc2) into a CL5 error (rc), but then this translation is lost when rc (CL5) is overwriten by rc2 (DBI).

Reply:
You are right: I wanted to avoid overwriting rc if there was a previous error but I mixed up things!
the 1911 line should be moved to replace 1913 one ...

Reply:
Now fixed.